Try pnpm (take two)

sarayourfriend opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
I'll force push this onto `try/pnpm` in https://github.com/WordPress/gutenberg/pull/37324 once this is working with CI so we can preserve the original PR discussion.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->
- [ ] I've updated related schemas if appropriate. <!-- Reference: https://github.com/WordPress/gutenberg/tree/trunk/schemas -->

sarayourfriend commented:
The `Pull request automation` job is going to continue failing until the new action I've added is on `trunk`. Not sure there's a way around that for this initial PR.

sarayourfriend commented:
Failing on installation due to some patches not applying correctly. Either we need to hoist the packages we're patching _or_ I think we could also just change the patch diffs to point to the `.pnpm` prefix that pnpm uses. I'll experiment with both, first trying to reproduce the issue locally.

sarayourfriend commented:
I think I need to bust the cache somehow 🤔 Seems like the cache calculation should include `.npmrc` too 😅

sarayourfriend commented:
That worked!

Now there seem to be some TypeScript resolution issues. Cursory searching seems to indicate TypeScript should "just work" with pnpm but clearly there's some issue with our setup here.

I also can't reproduce the issue locally yet, so that's probably the first step.

Will report back anything I find 🙂

sarayourfriend commented:
Aha! Looking at the first couple types that aren't being resolved it seems like they're indeclared development dependencies! `pnpm` is already doing it's job for us and pointing out assumptions we've made in the dependency chain! Marvelous 🤩

sarayourfriend committed:
Run pnpm import and squash other changes to ease rebasing

sarayourfriend commented:
Had to squash changes to make rebasing easier when `package-lock.json` changes on `trunk`. Sorry if this makes the history and comments confusing!

sarayourfriend commented:
The plugin-zip building script needs to be updated to use `pnpm`. `npm cache verify` doesn't need to happen anymore I don't believe as `pnpm` will automatically verify the cache when running `pnpm install` (maybe??). I'll do some more research on this and try to discover what the right way of doing that with `pnpm` is.

sarayourfriend committed:
Remove package-lock.json linting

sarayourfriend committed:
Declare dependency on csstype

sarayourfriend committed:
Fix wonky @types/react resolution due to react-native having different versions declared

sarayourfriend committed:
Try fixing react type resolution with hoisting

sarayourfriend committed:
Try a combination of hositing and tsconfig paths

sarayourfriend committed:
Add explicit dev dependency on @types/node

sarayourfriend committed:
Add explicit jest-circus dependency

sarayourfriend committed:
Remove lockfile linting step

sarayourfriend committed:
Modify performance test script for pnpm

sarayourfriend commented:
This will on some level be blocked by until this PR is merged for the compressed-size-check action: https://github.com/preactjs/compressed-size-action/pull/68

sarayourfriend committed:
Use `pnpm list` to list installed packages

sarayourfriend committed:
Use `pnpm` to install and build plugin zip

sarayourfriend committed:
Fix running native package scripts

sarayourfriend committed:
Fix build script

sarayourfriend committed:
Try hoisting within a package

sarayourfriend commented:
Using `pnpm` to run the build script seems to have moved the progress on this PR backwards 😰 I need to put this down for now but will revisit it next week again (when there are more merge conflicts to resolve, I am sure!)

kevin940726 commented on .github/workflows/create-block.yml:
```diff
@@ -29,14 +29,11 @@ jobs:
               with:
-                  node-version: ${{ matrix.node }}
-                  cache: npm
+                  node-version: ${{ matrix-node }}
```
I think this should still be `matrix.node`?

sarayourfriend replied:
Whoops, yes it should be!

kevin940726 commented on .github/workflows/flaky-tests.yml:
```diff
@@ -14,15 +14,9 @@ jobs:
-              uses: actions/setup-node@38d90ce44d5275ad62cc48384b3d8a58c500bb5f # v2.2.2
+            - uses: ./.github/action/setup-node-env
               with:
-                  node-version: 14
```
Better to keep this?

sarayourfriend replied:
The default is set to 14. I only kept passing it to the action in places that were still using the matrix strategy for it.

kevin940726 commented on .github/action/setup-node-env/action.yml:
```diff
@@ -0,0 +1,53 @@
+        - name: Install
+          if: inputs.install
+          shell: bash
+          run: "pnpm install ${{ inputs.frozen-lockfile && '--frozen-lockfile' || '' }}"
```
Looking at the [doc](https://pnpm.io/cli/install#--frozen-lockfile), seems like we don't have to pass `--frozen-lockfile` to true in CI since it's the default.

sarayourfriend replied:
Interesting. I don't really understand the "For CI: true" remark. How can it tell if it's in a CI environment? This isn't described anywhere in the documentation as far as I could tell... might have to dig into the code to find that out when I come back to this.

I do think you're right in questioning why we would need it to ever _not_ install with a frozen-lockfile on CI though. Maybe this PR _is_ the right time to investigate that? 🤔

kevin940726 replied:
I believe most programs depend on the env variable `process.env.CI` to exist in CI environment. Most CI platforms, including GH Actions, set this variable.

kevin940726 commented on .github/action/setup-node-env/action.yml:
```diff
@@ -0,0 +1,53 @@
+        required: false
+        description: Optional working directory to run installation in.
+    frozen-lockfile:
+        default: false
```
Why do we want to default it to false?

sarayourfriend replied:
I'm not sure! There were places that were doing just `npm install` rather than `npm ci` and I didn't want to change their behavior in this PR.

sarayourfriend replied:
In https://github.com/WordPress/gutenberg/pull/38624/commits/bd1d499c05c0a01cae5b3c157ca549130f551102 I changed it to just do a regular frozen-lockfile install everywhere relying on the default behavior in CI you mentioned. I can't think of a reason why CI should need to use a non-frozen install anywhere anyway.

kevin940726 commented on .github/action/setup-node-env/action.yml:
```diff
@@ -0,0 +1,53 @@
+        default: false
+        description: Whether to install dependencies using `--frozen-lockfile`.
+    node-version:
+        default: 14
```
I think this is nice that we don't have to repeat passing `node-version` in every workflow if it's just running the default version. We can just update this file when upgrading node to a major version.

ocean90 replied:
I guess this could also be changed to `node-version-file: '.nvmrc'` so that https://github.com/WordPress/gutenberg/blob/trunk/.nvmrc is the only place to change the version.

https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file

sarayourfriend replied:
Nice find! That sounds like a good option 🎉 We'd need to keep around `node-version` still for the workflows that use a node version matrix but referring to the `.nvmrc` file for the default is 💯

gziolo replied:
`.nvmrc ` plays an important role for branches where we apply fixes to older WordPress major releases like `wp/5.1`.

sarayourfriend replied:
Okay in https://github.com/WordPress/gutenberg/pull/38624/commits/bd1d499c05c0a01cae5b3c157ca549130f551102 I tried to set it up to just use `.nvmrc` by default unless a specific `node-version` is passed in (to preserve the ability for workflows to use the node version matrix, if that's still necessary). Hopefully it works, I'm not too strong on GitHub action expressions but should only need a little tweaking if it doesn't work right off the bat.

kevin940726 commented on package.json:
```diff
@@ -216,87 +220,83 @@
-		"check-local-changes": "( git diff -U0 | xargs -0 node bin/process-git-diff ) || ( echo \"There are local uncommitted changes after one or both of 'npm install' or 'npm run docs:build'!\" && git diff --exit-code && exit 1 );",
-		"dev": "npm run build:packages && concurrently \"wp-scripts start\" \"npm run dev:packages\"",
+		"precheck-local-changes": "pnpm docs:build",
+		"check-local-changes": "( git diff -U0 | xargs -0 node bin/process-git-diff ) || ( echo \"There are local uncommitted changes after one or both of 'npm install' or 'pnpm docs:build'!\" && git diff --exit-code && exit 1 );",
```
```suggestion
		"check-local-changes": "( git diff -U0 | xargs -0 node bin/process-git-diff ) || ( echo \"There are local uncommitted changes after one or both of 'pnpm install' or 'pnpm docs:build'!\" && git diff --exit-code && exit 1 );",
```

sarayourfriend replied:
Done in https://github.com/WordPress/gutenberg/pull/38624/commits/618ae4ff75845f0d106830dfe0eaa97237ecccbb

kevin940726 commented on packages/scripts/scripts/check-licenses.js:
```diff
@@ -167,14 +167,13 @@ const checkLicense = ( allowedLicense, licenseType ) => {
 };
 
-// Use `npm ls` to grab a list of all the packages.
+// Use `pnpm ls` to grab a list of all the packages.
```
```suggestion
// Use `pnpm list` to grab a list of all the packages.
```

sarayourfriend replied:
Done in https://github.com/WordPress/gutenberg/pull/38624/commits/618ae4ff75845f0d106830dfe0eaa97237ecccbb

kevin940726 reviewed (commented):
Thank you for starting this! Spotted some tiny errors.

Will be happy to try it out once this is ready 👍

sarayourfriend committed:
Improve setup node action and use better defaults

sarayourfriend committed:
Correct documentation/output

sarayourfriend committed:
Try hositing inside the package that actually uses it

sarayourfriend committed:
Try shamefully-hoist

sarayourfriend committed:
Fix workspace package references

sarayourfriend commented:
I just realized I'd been fighting dependency resolution issues through hoisting that are meant to be resolved through cross-workspace-package dependencies... Turns out `file` is meant to be replaced with `workspace`.

sarayourfriend committed:
Try using link: for workspace root dependencies

https://github.com/pnpm/pnpm/issues/3642#issuecomment-948957718

sarayourfriend committed:
Patch eslint module resolution

See https://github.com/microsoft/rushstack/tree/master/eslint/eslint-patch
README for explanation why this is needed in a monorepo setting.

sarayourfriend committed:
Hoist stylelint dependencies

sarayourfriend commented:
~Turns out eslint and stylelint are both very special and refuse to resolve plugin modules in a modern way. `eslint` has a widely used patch for this behavior but stylelint does not, so the latter requires us to hoist its dependencies.~

It turns out the patch will only work for a top-level eslint configuration. The nested `react-native-editor` configuration won't be helped by the patch and requires hoisting of the eslint-plugin dependencies anyway. I removed the patch and added a lengthy explanation for why we're hoisting those packages.

In general I think it'd be a good idea to continue documenting hoists. I _feel_ like we can avoid just hoisting everything (which has it's own set of problems anyway) and everything that is documented and hoisted should theoretically be able to eventually be removed with the right dependency declarations.

Of course, that's with the exception of stylelint, eslint and anything used by expo, because those tools all have their own non-modern module resolution strategies that cause problems with `pnpm`s strictness.

Consumers of the `wp-scripts` package outside the monorepo will not be effected by this as the dependencies will install at the top level of their repositories (unless they're also using a monorepo, I suppose, but then only if they use `pnpm` will they run into it).

Anyway, it's kind of a mess but unavoidable at this point as far as I can tell.

sarayourfriend committed:
Break down and just hoist stuff

sarayourfriend committed:
Hoist namespaced plugins

sarayourfriend committed:
Try workaround for npm-run-all not supporting pnpm

See https://github.com/mysticatea/npm-run-all/issues/201

sarayourfriend committed:
Add missing top-level dev dependency

sarayourfriend commented:
Regarding the current state of the unit tests: I think they might be flagging a potentially problematic (not generally, just for `pnpm`'s strictness) assumption in the WebpackExtractDependenciesPlugin.

The fact that the e2e tests are passing consistently is exciting though!

sarayourfriend commented:
Closing this draft. If anyone wants to pick this up once React Native is supported by pnpm then you'll have some starting points here though I would be things have changed in the mean time. The new npm versions might not make this change worthwhile anyway.

manzoorwanijk commented:
React Native [now supports symlinks](https://reactnative.dev/blog/2023/06/21/0.72-metro-package-exports-symlinks) 🎉 

Do you think we can go for take three?
